io-engine/src/grpc/v1/nexus.rs

```diff
@@ -1147,4 +1148,47 @@ impl NexusRpc for NexusService {
         })
         .await
     }
+
+    async fn list_rebuild_history(
+        &self,
+        request: Request<ListRebuildHistoryRequest>,
+    ) -> GrpcResult<RebuildHistoryMapResponse> {
+        let args = request.into_inner();
+        let count = args.count.unwrap_or_default();
+        let rx = rpc_submit::<_, _, nexus::Error>(async move {
+            let mut history_map: HashMap<String, RebuildHistoryResponse> =
+                HashMap::new();
+            for nexus in nexus::nexus_iter() {
+                let mut history_list: VecDeque<RebuildHistoryRecord> =
+                    VecDeque::new();
+                let nexus_uuid = nexus.uuid().to_string();
+                for record in
+                    nexus_lookup(&nexus_uuid)?.rebuild_history().iter()
+                {
+                    // We will return all of the rebuild records for nexus if
+                    // count is 0,Apply only non zero
+                    // limit.
+                    if history_list.len() == count.encoded_len() && count != 0 {
+                        history_list.pop_front();
+                    };
+                    history_list.push_back(RebuildHistoryRecord::from(record));
```

Comment:
You don't need to have a mut hashmap, you can write all this in completely functional way:
```
let hist_map: HashMap<_, _> = nexus::nexus_iter()
    .map(|nex| {
        let uuid = nex.uuid().to_string();

        // get and modify history list here as needed
        let hist = nex.rebuild_history();
        let s = if count == 0 || count > hist.len() {
            0
        } else {
            hist.len() - count
        };
        let hist = hist[s..].to_owned();

        // return a tuple of (uuid, RebuildHistoryResponse)
        (
            uuid.clone(),
            RebuildHistoryResponse {
                nexus: uuid.clone(),
                records: hist.into(),
            },
        )
    })
    .collect();  // Collecting an iterator of tuples will work for a HashMap
```

Reply:
Incorporated

Reply:
hmm you can also use rev and take to simplify the limit:
and probably don't need to clone uuid the second time

```rust
        let hist = nex.rebuild_history();
        let hist = hist.into_iter().rev().take(count).collect::<Vec<_>>();

        // return a tuple of (uuid, RebuildHistoryResponse)
        (
            uuid.clone(),
            RebuildHistoryResponse {
                nexus: uuid,
                records: hist.into(),
            },
        )
```

Reply:
removed the clone already. Ill try rev

Reply:
Incorporated